### PR TITLE
Plot frame

### LINKED ISSUE #268: Setting the coordinate reference system
When plotting images with `add_image()` it doesn't show up in the conventional orientation, but upside down (same when using `ImageWidget()`.

```
face = np.array([
   [0.6, .5, .5,  0.6, 0.5, 0.5, 0.6,], # row 0
   [0.6, 0.4, .5, 0.6, 0.4, 0.5, 0.6,], # row 0.6
   [0.6, 0.6, 0.6, 0.6, 0.6, 0.6, 0.6,], # row 2
   [0.6, 0.6, 0.6, 0.4, 0.6, 0.6, 0.6,], # row 3
   [0.6, 0.4, 0.6, 0.6, 0.6, 0.4, 0.6,], # row 4
   [0.6, 0.6, 0.4, 0.4, 0.4, 0.6, 0.6,], # row 5
   [0.5, 0.6, 0.6, 0.6, 0.6, 0.6, 0.5,], # row 6
]);
```
Matplotlib  `plt.imshow(face, cmap='gray')` output:

mpl_imshow

Fastplotlib:
```
import fastplotlib as fpl
plot = fpl.Plot()
plot.add_image(data=face, cmap='gray')
plot.show()
```
add_image

Obviously there are conventions at play here with yaxis being switched being the norm for images.

### LINKED ISSUE #310: make a `PlotFrame` class for `Toolbar` and `sidecar` functionality
TODO once #300 is merged

### LINKED ISSUE #342: make `sidecar=False` the default?
@clewis7 @EricThomson your thoughts would be great. Most libraries put plotting outputs right after the cell so maybe we should follow that as the default? It also avoids things like this: https://github.com/fastplotlib/fastplotlib/issues/337

### PR DESCRIPTION
**Pull request by kushalkolar**
closes #310 closes #342 closes #268

This PR adds a `Frame` class that manages how plots are shown. `Plot` and `Gridplot`, and any future flexilbe gridplot, inherit from `Frame`. 

### Toolbars
Refactored ipywidget Toolbar that is common between Plot, Gridplot and any future gridplots

### Output Contexts
These are classes that "hold" the output of a Plot/Gridplot. They determine how the canvas, toolbar, and any additional widgets are shown. When `Plot.show()` is called, they return an instance of one of the following output contexts based on the curent plot canvas.

  - `JupyterOutputContext`
    - Inherits from `ipywidgets.VBox` and contains the canvas, toolbar, and any other widgets. Manages sidecar if desired.
  - `QOutputContext`
    - Inherits from `QWidget`, contains the canvas, qt toolbar, and any other widgets.
  - If the canvas is not jupyter or qt (ex. if it's glfw or offscreen like in our test suite), the output context is just the plot canvas itself. 

This cleans up the messy `show()` calls we previously had everywhere, and cleans up other stuff too. Now the output context manages closing plots too. 

BTW, the `QOutputContext` can also be used in notebooks with if you execute `%gui qt` before creating plots :smile:

### CODE REVIEW

**Comment by clewis7** [comment]
#310

**Comment by kushalkolar** [comment]
This is working now with `JupyterOutputContext`! Also fixed the wonky issue with auto-scale and center scene when you had a single controller controlling multiple cameras for diff subplots. 

Now onto Qt output context and toolbar.

**Comment by kushalkolar** [comment]
Everything works on my end, I tested every combination with ipywidgets, sidecar/no sidecar, Qt, and with ImageWidget. 

Right now the toolbar implementations are quite repetitive, we can make that nicer in a future PR after we have some time to think about it. Someone must have made a common interface between ipywidgets and Qt that we can use. 

https://github.com/fastplotlib/fastplotlib/assets/9403332/462d7813-850a-492c-95fc-12c691e2fbb9

File: fastplotlib/layouts/_frame/_frame.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        if add_widgets is None:
+            add_widgets = list()
+
+        # flip y axis if ImageGraphics are present
+        if hasattr(self, "_subplots"):
+            for subplot in self:
+                for g in subplot.graphics:
+                    if isinstance(g, ImageGraphic):
+                        subplot.camera.local.scale_y = -1
+                        break
```
Comment (clewis7) [inline fastplotlib/layouts/_frame/_frame.py:144]: closes #268

File: fastplotlib/layouts/_frame/_frame.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        # show was already called, return existing output context
+        if self._output is not None:
+            return self._output
+
+        self.start_render()
+
+        if sidecar_kwargs is None:
+            sidecar_kwargs = dict()
+
+        if add_widgets is None:
```
Comment (clewis7) [inline fastplotlib/layouts/_frame/_frame.py:135]: I assume these are any additional widgets someone would want VBoxed with the plot?

File: fastplotlib/layouts/_frame/_ipywidget_toolbar.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+        # set initial values for some buttons
+        self._maintain_aspect_button.value = self.current_subplot.camera.maintain_aspect
+
+        if copysign(1, self.current_subplot.camera.local.scale_y) == -1:
+            self._y_direction_button.icon = "arrow-down"
+        else:
+            self._y_direction_button.icon = "arrow-up"
+
+        HBox.__init__(self, widgets)
```
Comment (clewis7) [inline fastplotlib/layouts/_frame/_ipywidget_toolbar.py:137]: why this notation?

File: fastplotlib/layouts/_frame/_ipywidget_toolbar.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+)
+
+from ...graphics.selectors import PolygonSelector
+from ._toolbar import ToolBar
+
+
+class IpywidgetToolBar(HBox, ToolBar):
+    def __init__(self, plot):
+        """
+        Basic toolbar for a GridPlot instance.
```
Comment (clewis7) [inline fastplotlib/layouts/_frame/_ipywidget_toolbar.py:29]: for GridPlot or Plot...correct?

File: fastplotlib/layouts/_frame/_jupyter_output.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                f"add_widgets must be list of ipywidgets, you have passed:\n{add_widgets}"
+            )
+
+        self.use_sidecar = use_sidecar
+
+        if not make_toolbar:  # just stack canvas and the additional widgets, if any
+            self.output = (frame.canvas, *add_widgets)
+
+        if make_toolbar:  # make toolbar and stack canvas, toolbar, add_widgets
+            self.toolbar = IpywidgetToolBar(frame)
```
Comment (clewis7) [inline fastplotlib/layouts/_frame/_jupyter_output.py:53]: what if ImageWidget? where is logic for if plot type is imagewidget?

**Comment by clewis7** [comment]
did you update the requirements file?

File: fastplotlib/layouts/_frame/_frame.py (hunk 1 above)
Comment (clewis7) [reply fastplotlib/layouts/_frame/_frame.py:144]: One thing I notice is that this logic is not implemented for imagewidget

**Comment by clewis7** [comment]
Another thing I see is that when I have two subplots in an imagewidget and the camera y_scales are opposite that when I change click between subplots the flip button is not changing in accordance with the camera y_scale...just stays the same

File: fastplotlib/layouts/_frame/_frame.py (hunk 1 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_frame.py:144]: ImageWidget uses a Gridplot so it's not necessary

File: fastplotlib/layouts/_frame/_frame.py (hunk 2 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_frame.py:135]: Yes

File: fastplotlib/layouts/_frame/_ipywidget_toolbar.py (hunk 3 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_ipywidget_toolbar.py:137]: Instantiate the subclass, I could do a super instead too

**Comment by kushalkolar** [comment]
> did you update the requirements file?

No dependency changes here.

**Comment by clewis7** [comment]
> > did you update the requirements file?
> 
> No dependency changes here.

not even if someone only wants qt? they would need to have pyqt6 installed right?

**Comment by kushalkolar** [comment]
> > > did you update the requirements file?
> > 
> > 
> > No dependency changes here.
> 
> not even if someone only wants qt? they would need to have pyqt6 installed right?

Qt land is a bit complicated, so they will have to install Qt themselves separately. We can do toggles later, because to do this properly we need to give users the option of PySide etc and decide how broad we want this support.

File: fastplotlib/layouts/_frame/_ipywidget_toolbar.py (hunk 4 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_ipywidget_toolbar.py:29]: yes!

File: fastplotlib/layouts/_frame/_ipywidget_toolbar.py (hunk 3 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_ipywidget_toolbar.py:137]: changed to super

File: fastplotlib/layouts/_frame/_jupyter_output.py (hunk 5 above)
Comment (kushalkolar) [reply fastplotlib/layouts/_frame/_jupyter_output.py:53]: `ImageWidget` just uses the `Gridplot` context:

```python
    def show(self, toolbar: bool = True, sidecar: bool = False, sidecar_kwargs: dict = None):
        """
        Show the widget

        Returns
        -------
        OutputContext
        """
        if self.gridplot.canvas.__class__.__name__ == "JupyterWgpuCanvas":
            self._image_widget_toolbar = IpywidgetImageWidgetToolbar(self)

        elif self.gridplot.canvas.__class__.__name__ == "QWgpuCanvas":
            self._image_widget_toolbar = QToolbarImageWidget(self)

        return self.gridplot.show(
            toolbar=toolbar,
            sidecar=sidecar,
            sidecar_kwargs=sidecar_kwargs,
            add_widgets=[self._image_widget_toolbar]
        )
```

**Comment by kushalkolar** [comment]
> > > Another thing I see is that when I have two subplots in an imagewidget and the camera y_scales are opposite that when I change click between subplots the flip button is not changing in accordance with the camera y_scale...just stays the same
> > 
> > 
> > In ipywidgets or Qt?
> 
> ipwidgets

I think now I see what you mean, it is wonky! So this is because in ImageWidget all cameras are under one controller. I did not decide how to deal with flipping when controllers are diff. What I will do is if the flip button is clicked, it will flip the camera for the current subplot but also any camera that has the same controller (similar to what I did for panzoom and maintain_aspect).

**Pull request merged**